portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -23,14 +33,93 @@
 public class DataServlet extends HttpServlet {
-
+  final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("UserData");
+    ArrayList<UserData> data = new ArrayList<UserData>();	
+    PreparedQuery results = datastore.prepare(query);
+
+    Date currentDate = new Date();  
+    
+    for (Entity entity : results.asIterable()) { 
+      long id = entity.getKey().getId();
+      String name = (String) entity.getProperty("name");
+      String email = (String) entity.getProperty("email");
+      String comment = (String) entity.getProperty("comment");
+      Date date = (Date) entity.getProperty("date");
+      UserData newData = new UserData(id, name, email, comment, date);
+      data.add(newData);
+    }
+
+    Gson gson = new Gson();
+    int numComments = filterCommentsByNum(request, "numValue", 4);
+    
     response.setContentType("application/json;");
-    // response.getWriter().println("Hello Chisom!");
-    String jsonComments = "{\"Femi\": \"Nice page Chisom\", \"Dera\": \"Try to make the web page reponsive\", \"Kamsi\": \"Kudos\"}";
-    response.getWriter().println(jsonComments);
+    if (numComments > data.size()){ 
+    response.getWriter().println(gson.toJson(data.subList(0, data.size())));
+    }
+    else{
+    response.getWriter().println(gson.toJson(data.subList(0, numComments )));
+    }
+  }
+
+  public static class UserData{
+      long id;
+      String name;
+      String email;
+      String comment;
+      Date date;
+
+      UserData(long id, String name, String email, String comment, Date date){
+          this.id = id;
+          this.name = name;
+          this.email = email;
+          this.comment = comment;
+          this.date = date;
+      }
   }
 
+  @Override
+   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+       Date date = new Date();  
+       String name = getParameter(request, "name", "");
+       String email = getParameter(request, "email", "");
+       String comment = getParameter(request, "comment", "");
+
+        if(name.isBlank())
+         throw new IllegalArgumentException("Name not entered.");
+
+        if (comment.isBlank()) 
+         throw new IllegalArgumentException("Comment field empty.");
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Entity taskEntity = new Entity("UserData");
+    	taskEntity.setProperty("name", name);
+   	 	taskEntity.setProperty("email", email);
+        taskEntity.setProperty("comment", comment);
+   	 	taskEntity.setProperty("date", date);
+        datastore.put(taskEntity);
+
+        // Redirect back to the HTML page.
+    	response.sendRedirect("/index.html");
+    }
+
+    private int filterCommentsByNum(HttpServletRequest request,  String name, int defaultValue) {
```

Comment:
let's consider the naming of this function: is this function filtering comments by some number? or is it retrieving an arbitrary parameter which may have nothing to do with comments or filtering?

Reply:
I renamed it getNumComments